### PR TITLE
core[minor]: Nc/dec26/runnable stream

### PR DESCRIPTION
**Pull request by nfcampos**
Fixes # (issue)

### CODE REVIEW

File: langchain-core/src/runnables/base.ts [hunk 1]
```diff
@@ -1537,22 +1588,29 @@ export class RunnableLambda<RunInput, RunOutput> extends Runnable<
...
+      if (config?.recursionLimit === 0) {
+        throw new Error("Recursion limit reached.");
+      }
       output = await output.invoke(
         input,
-        this._patchConfig(config, runManager?.getChild())
+        this._patchConfig(
+          config,
+          runManager?.getChild(),
+          (config?.recursionLimit ?? 25) - 1
```
Comment (jacoblee93) [inline langchain-core/src/runnables/base.ts:1604]: Factor out magic number 25 somewhere else?

File: langchain-core/src/runnables/passthrough.ts [hunk 2]
```diff
@@ -31,6 +33,66 @@ export class RunnableAssign<
...
     } as RunOutput;
   }
+
+  async *_transform(
+    generator: AsyncGenerator<RunInput>,
+    runManager?: CallbackManagerForChainRun,
+    options?: Partial<RunnableConfig>
+  ): AsyncGenerator<RunOutput> {
+    // collect mapper keys
+    const mapper_keys = this.mapper.getStepsKeys();
```
Comment (jacoblee93) [inline langchain-core/src/runnables/passthrough.ts:43]: Prefer camel case

File: langchain-core/src/utils/stream.ts [hunk 3]
```diff
@@ -98,3 +98,63 @@ export class IterableReadableStream<T>
     });
   }
 }
+
+export function atee<T>(
```
Comment (jacoblee93) [inline langchain-core/src/utils/stream.ts:102]: Just `tee` is better?

File: langchain-core/src/runnables/passthrough.ts [hunk 4]
```diff
@@ -31,6 +33,66 @@ export class RunnableAssign<
...
+    // create two input gens, one for the mapper, one for the input
+    const [forPassthrough, forMapper] = atee(generator, 2);
+    // create mapper output gen
+    const mapperOutput = this.mapper.transform(
```
Comment (jacoblee93) [inline langchain-core/src/runnables/passthrough.ts:47]: What's the purpose of the tee? Can we not just:

```typescript
      const filtered = Object.fromEntries(
        Object.entries(chunk).filter(([key]) => !mapper_keys.includes(key))
      );
      if (Object.keys(filtered).length > 0) {
        yield filtered as unknown as RunOutput;
      }
```
below along with the other chunks?

File: langchain-core/src/utils/stream.ts (hunk 3 above)
Comment (nfcampos) [reply langchain-core/src/utils/stream.ts:102]: Tee would be for non-async generators

File: langchain-core/src/runnables/base.ts [hunk 5]
```diff
@@ -433,30 +428,24 @@ export abstract class Runnable<
...
+      { input: "" },
+      undefined,
+      options?.runType,
+      undefined,
+      undefined,
+      options?.runName
+    );
     async function* wrapInputForTracing() {
       for await (const chunk of inputGenerator) {
-        if (!runManager) {
```
Comment (jacoblee93) [inline langchain-core/src/runnables/base.ts:440]: Ah nice that this is fixed

File: langchain-core/src/runnables/passthrough.ts (hunk 4 above)
Comment (nfcampos) [reply langchain-core/src/runnables/passthrough.ts:47]: We can't consume the same generator twice

File: langchain-core/src/utils/stream.ts [hunk 6]
```diff
@@ -98,3 +98,63 @@ export class IterableReadableStream<T>
...
+export function concat<
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  T extends Array<any> | string | number | Record<string, any> | any
+>(first: T, second: T): T {
+  if (Array.isArray(first) && Array.isArray(second)) {
+    return first.concat(second) as T;
+  } else if (typeof first === "string" && typeof second === "string") {
+    return (first + second) as T;
+  } else if (typeof first === "number" && typeof second === "number") {
+    return (first + second) as T;
```
Comment (jacoblee93) [inline langchain-core/src/utils/stream.ts:136]: This seems like an odd thing to do

Comment (nfcampos) [reply langchain-core/src/utils/stream.ts:136]: This is what we do in py

File: langchain-core/src/runnables/passthrough.ts (hunk 4 above)
Comment (jacoblee93) [reply langchain-core/src/runnables/passthrough.ts:47]: Oh I get it, never mind.

**Comment by jacoblee93** [comment]
Thanks so much for this!

File: langchain-core/src/utils/stream.ts (hunk 6 above)
Comment (jacoblee93) [reply langchain-core/src/utils/stream.ts:136]: Streaming number would be a bit weird in general I guess

**Pull request merged**